server/aesdsocket.c

```diff
@@ -14,140 +14,288 @@
 if(ret<0){
     syslog(LOG_USER, "Binding not done.");
+    perror("bind");
+    return(-1);
 
 }
 freeaddrinfo(server_info);
-
+if(deamon){
+    int ret=daemon(-1,-1);
+    if(ret==-1){
+        perror("daemon");
+    }
+}
+struct thread_data* thread;
+ret=pthread_mutex_init(&mutex,NULL);
+SLIST_INIT(&head);
 syslog(LOG_USER, "Listening");
 listen(sockfd,BACKLOG);
-while(1){  
-    size_recived=0;
-    size_read=0;
-    syslog(LOG_USER, "Entered");  
+ret=set_time();
+if(ret!=0){
+    perror("setitimer");
+}
+syslog(LOG_USER, "Error4");
+while(!interrupted){ 
     new_sockfd=accept(sockfd,&client_addr,&addr_size);
-    syslog(LOG_USER, "Here1");  
     if(new_sockfd==-1){
         syslog(LOG_USER, "Error while accepting");
         perror("accept");
         close(sockfd);
         return(-1);
     }
+    struct thread_data* client_data=malloc(sizeof(struct thread_data));
+    client_data->socketfd=new_sockfd;
+    client_data->client=&client_addr;
+    client_data->connection_complete_success=false;
     syslog(LOG_USER, "Accepted Connection");
-    inet_ntop(AF_INET,&client_addr,clientIP,sizeof(clientIP));
-    bool recv_complete=false;
-    while(!recv_complete){
-        syslog(LOG_USER, "Inside recv loop");
-        s_recv=recv(new_sockfd,(buf+current_size),BUFFER_SIZE,0);
-        if(s_recv==-1){
-            syslog(LOG_USER, "Error while recieving");
-            return(-1);
-        }
-        size_recived+=s_recv;
-        if(size_recived>0 && buf[size_recived-1]=='\n'){
-            recv_complete=true;
-            syslog(LOG_USER, "Recieving complete");
-        }
-        current_size+=BUFFER_SIZE;
-    }
-    file_fd=open(FILE_PATH, O_RDWR | O_CREAT | O_APPEND ,0644);
-    if(file_fd == -1){
-        syslog(LOG_USER, "Unable to open file to read, Check the permission");
-        perror("open");
-        return(-1);
-    }     
-          while(nr=write(file_fd, &buf, size_recived)){
-        if(nr == -1){
-            syslog(LOG_USER, "Writing to file not Successfull");  
-            return(-1);
-            }     
-          }
+syslog(LOG_USER, "Error5");
+    ret=pthread_create(&(client_data->thread_id),NULL,thread_function,client_data);
+    //Check if creation of thread was successfull
+    if(ret!=0){
+        return (-1);
+    }  
+    SLIST_INSERT_HEAD(&head,client_data,entries);
+    if(client_data->connection_complete_success){
+        close(client_data->socketfd);
+        pthread_join(client_data->thread_id,NULL);
+        SLIST_REMOVE(&head,client_data, thread_data, entries);
 
-    char message[BUFFER_SIZE];
-    lseek(file_fd,0,SEEK_SET);
-    while(nr=read(file_fd,message,BUFFER_SIZE)!=0){
-        if(nr == -1){
-            syslog(LOG_USER, "Reading from file not Successfull");   
-            return(-1);
-        }  
-        s_send=send(new_sockfd,message,nr,0);
-        if(s_send<0){
-           syslog(LOG_USER, "Sending failed");  
-        }
-    }   
-        
+    }
 }
+close(file_fd);
```

Comment:
In case of SIGINT does your code handle clean up of remaining pthreads?

Reply:
The while loop inside the signal handler is supposed to do that. Hope that answers the question.